ARROW-14161: [C++][Docs] Improve Parquet C++ docs (#14018)
apache/arrow: docs/source/cpp/parquet.rst

C++ implementation is part of the Apache Arrow project and benefits
from tight integration with the Arrow C++ classes and facilities.

Reading Parquet files
=====================

The :class:`arrow::FileReader` class reads data for an entire

> file or row group into an :class:`::arrow::Table`.

Reviewer: This happens a few times throughout the article. Why `::arrow::Table` instead of `arrow::Table`, or even including the Arrow namespace in the first place and just using `Table`?

Author: I'll delete the leading `::`. We are in the `parquet` namespace here, not `arrow`, and I think it would be confusing to include both, so I'll keep the `arrow::` prefixes.

The :class:`StreamReader` and :class:`StreamWriter` classes allow for
data to be written using a C++ input/output streams approach to

> read/write fields column by column and row by row.  This approach is

Reviewer: Is this value by value, or a choice between reading full rows or full columns?

Author: value by value, IIUC. I didn't touch this part of this docs since this part of the interface isn't that important.

offered for ease of use and type-safety.  It is of course also useful
when data must be streamed as files are read and written
incrementally.

Please note that the performance of the :class:`StreamReader` and
:class:`StreamWriter` classes will not be as good due to the type
checking and the fact that column values are processed one at a time.

FileReader
----------

The Parquet :class:`arrow::FileReader` requires a
:class:`::arrow::io::RandomAccessFile` instance representing the input
file.

.. literalinclude:: ../../../cpp/examples/arrow/parquet_read_write.cc
   :language: cpp
   :start-after: arrow::Status ReadFullFile(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 9-10,14
   :dedent: 2

> Finer-grained options are available through the

Reviewer: I'd suggest putting a sub-header here -- I was expecting linearity, and had to double-take when I realized the code example following this prose does not follow from the one above, but is instead another path to file reading.

Author: Well this is a user guide, not a tutorial, so I think the linearity expectation isn't as strong. Each snippet is meant to be a (mostly) standalone example.

Reviewer: The linearity expectation isn't as strong, but I wasn't quite able to realize each snippet was standalone without a fresh header.

Author: I added more to that first paragraph so it's clearer that first code snippet is a full example of reading a whole file.

Reviewer: Great!

> :class:`arrow::FileReaderBuilder` helper class, and the :class:`ReaderProperties`

Reviewer: ```suggestion
:class:`arrow::FileReaderBuilder` helper class, when paired with the :class:`ReaderProperties`
```

It appears to me that you use the property classes in tandem with the `FileReaderBuilder`, so it seems worth being explicit about that.

and :class:`ArrowReaderProperties` classes.

> For reading as a stream of batches, use the :func:`arrow::FileReader::GetRecordBatchReader`.

Reviewer: ```suggestion
For reading as a stream of batches, use the :class:`arrow::RecordBatchReader`, which you can get via :func:`arrow::FileReader::GetRecordBatchReader`.
```

It will use the batch size set in :class:`ArrowReaderProperties`.

.. literalinclude:: ../../../cpp/examples/arrow/parquet_read_write.cc
   :language: cpp
   :start-after: arrow::Status ReadInBatches(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 25
   :dedent: 2

Performance and Memory Efficiency
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> For remote filesystems, use read coalescing to reduce number of API calls:

Reviewer: ```suggestion
For remote filesystems, use read coalescing (pre-buffering) to reduce number of API calls:
```

.. code-block:: cpp

   auto reader_properties = parquet::ReaderProperties(pool);
   reader_properties.enable_buffered_stream();
   reader_properties.set_buffer_size(4096 * 4); // This is default value

The defaults are generally tuned towards good performance, but parallel column
decoding is off by default. Enable it in the constructor of :class:`ArrowReaderProperties`:

.. code-block:: cpp

   auto arrow_reader_props = parquet::ArrowReaderProperties(/*use_threads=*/true);

If memory efficiency is more important than performance, then:

> #. Do not turn on read coalescing (pre-buffering).

Reviewer: Could drop this pre-buffering if you add the one above.

Author: Well these are two distinct pieces of advice:
If you are using a remote filesystem: turn **on** prebuffering
If you care about reducing memory usage: turn **off** prebuffering

Reviewer: No, I meant the "(pre-buffering)", the literal text; that you could drop the bit in parentheses if you wanted, if you added it to the previous "read coalescing" introduction.

Author: It's a little repetitive, but I think I want to keep it, since the API itself is confusing. The inline docs say "read coalescing" (which is more descriptive of what's going on) but the actual option in the API is `pre_buffer`.

#. Read data in batches.
#. Turn off ``use_buffered_stream``.

In addition, if you know certain columns contain many repeated values, you can

> read them as dictionary encoded columns. This is enabled with the ``set_read_dictionary``

Author: Good question. I'll add a link to the glossary reference for now.

But I don't see any mention of dictionary encoded columns in the user guide yet. I'll create a Jira to revamp the Arrays section, and make sure to include encoded columns.

> read them as dictionary encoded columns. This is enabled with the ``set_read_dictionary``

Reviewer: Sounds good to me!

explicitly by using the :type:`EndRowGroup` stream modifier.

Exceptions are used to signal errors.  A :class:`ParquetException` is
thrown in the following circumstances:

* Attempt to write a field using an incorrect type.

* Attempt to write too many fields in a row.

* Attempt to skip a required field.

.. code-block:: cpp

   #include "arrow/io/file.h"
   #include "parquet/stream_writer.h"

   {
      std::shared_ptr<arrow::io::FileOutputStream> outfile;

      PARQUET_ASSIGN_OR_THROW(
         outfile,
         arrow::io::FileOutputStream::Open("test.parquet"));

      parquet::WriterProperties::Builder builder;
      std::shared_ptr<parquet::schema::GroupNode> schema;

      // Set up builder with required compression type etc.
      // Define schema.
      // ...

      parquet::StreamWriter os{
         parquet::ParquetFileWriter::Open(outfile, schema, builder.build())};

      // Loop over some data structure which provides the required
      // fields to be written and write each row.
      for (const auto& a : getArticles())
      {
         os << a.name() << a.price() << a.quantity() << parquet::EndRow;
      }
   }

> Writer properties

Author: Yeah I added the reader properties into the existing writer examples. I showcased setting compression to snappy and turning on saving the Arrow schema, which we recommend pretty much everyone does.

representing the input file. To read the whole file at once, 
use :func:`arrow::FileReader::ReadTable`:

.. literalinclude:: ../../../cpp/examples/arrow/parquet_read_write.cc
   :language: cpp
   :start-after: arrow::Status ReadFullFile(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 9-10,14
   :dedent: 2

Finer-grained options are available through the
:class:`arrow::FileReaderBuilder` helper class, which accepts the :class:`ReaderProperties`
and :class:`ArrowReaderProperties` classes.

For reading as a stream of batches, use the :func:`arrow::FileReader::GetRecordBatchReader`
method to retrieve a :class:`arrow::RecordBatchReader`. It will use the batch 
size set in :class:`ArrowReaderProperties`.

.. literalinclude:: ../../../cpp/examples/arrow/parquet_read_write.cc
   :language: cpp
   :start-after: arrow::Status ReadInBatches(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 25
   :dedent: 2

.. seealso::

   For reading multi-file datasets or pushing down filters to prune row groups,
   see :ref:`Tabular Datasets<cpp-dataset>`.

Performance and Memory Efficiency
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> For remote filesystems, use read coalescing (pre-buffering) to reduce number of API calls:

Reviewer: Do users know what pre-buffering is? I assume yes. I have a vague idea, at least.

.. code-block:: cpp

   auto arrow_reader_props = parquet::ArrowReaderProperties(/*use_threads=*/true);

If memory efficiency is more important than performance, then:

#. Do *not* turn on read coalescing (pre-buffering) in :class:`parquet::ArrowReaderProperties`.
#. Read data in batches using :func:`arrow::FileReader::GetRecordBatchReader`.
#. Turn on ``enable_buffered_stream`` in :class:`parquet::ReaderProperties`.

In addition, if you know certain columns contain many repeated values, you can
read them as :term:`dictionary encoded<dictionary-encoding>` columns. This is 
enabled with the ``set_read_dictionary`` setting on :class:`ArrowReaderProperties`. 
If the files were written with Arrow C++ and the ``store_schema`` was activated,
then the original Arrow schema will be automatically read and will override this
setting.

StreamReader
------------

The :class:`StreamReader` allows for Parquet files to be read using
standard C++ input operators which ensures type-safety.

Please note that types must match the schema exactly i.e. if the
schema field is an unsigned 16-bit integer then you must supply a

> uint16_t type.

Reviewer 2: ```suggestion
``uint16_t`` type.
```

Exceptions are used to signal errors.  A :class:`ParquetException` is
thrown in the following circumstances:

* Attempt to read field by supplying the incorrect type.

* Attempt to read beyond end of row.

* Attempt to read beyond end of file.

.. code-block:: cpp

   #include "arrow/io/file.h"
   #include "parquet/stream_reader.h"

   {
      std::shared_ptr<arrow::io::ReadableFile> infile;

      PARQUET_ASSIGN_OR_THROW(
         infile,
         arrow::io::ReadableFile::Open("test.parquet"));

>       parquet::StreamReader os{parquet::ParquetFileReader::Open(infile)};

Reviewer 2: I tend to implicitly interpret ``os`` as "output stream", which is misleading here. Just call it ``stream`` perhaps?

      std::string article;
      float price;
      uint32_t quantity;

      while ( !os.eof() )
      {
         os >> article >> price >> quantity >> parquet::EndRow;
         // ...
      }
   }

Writing Parquet files
=====================

WriteTable
----------

The :func:`arrow::WriteTable` function writes an entire
:class:`::arrow::Table` to an output file.

.. literalinclude:: ../../../cpp/examples/arrow/parquet_read_write.cc
   :language: cpp
   :start-after: arrow::Status WriteFullFile(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 19-21
   :dedent: 2

> .. warning::

Reviewer 2: We should keep warnings for dangerous things, experimental API markers, etc.

```suggestion
.. note::
```

      }
   }

.. _parquet-writer-properties:

Writer properties
-----------------

To configure how Parquet files are written, use the :class:`WriterProperties::Builder`:

.. code-block:: cpp

   #include "parquet/arrow/writer.h"
   #include "arrow/util/type_fwd.h"

   using parquet::WriterProperties;
   using parquet::ParquetVersion;
   using parquet::ParquetDataPageVersion;
   using arrow::Compression;

   std::shared_ptr<WriterProperties> props = WriterProperties::Builder()
      .max_row_group_length(64 * 1024)
      .created_by("My Application")
      .version(ParquetVersion::PARQUET_2_6)
      .data_page_version(ParquetDataPageVersion::V2)
      .compression(Compression::SNAPPY)
      .build();

The ``max_row_group_length`` sets an upper bound that takes precedent over the
``chunk_size`` passed in the write methods.

You can set the version of Parquet to write with ``version``, which determines
which logical types are available. In addition, you can set the data page version
with ``data_page_version``. It's V1 by default; setting to V2 will allow more
optimal compression (skipping compressing pages where there isn't a space 

> benefit), but not all readers support this data page version.

Reviewer: Not all readers that exist in general, even outside of Arrow, or do we have readers that fail?

Author: I think Kae is asking which readers don't support V2 pages; I think listing that support matrix is out-of-scope of this PR, but it would be nice to see.

Reviewer 2: Right, that should certainly happen at the level of the Parquet project, not Arrow.

Reviewer: Sorry for delay here, thought I responded -- I don't mean a full support matrix, I just mean saying whether or not "readers," as in classes in Arrow, all support V2. That is, is Arrow fully V2-accepting? If so, it may be worth making it more clear that external ones are the ones that're not fully compatible.

Compression is off by default, but to get the most out of Parquet, you should 
also choose a compression codec. You can choose one for the whole file or 
choose one for individual columns. If you choose a mix, the file-level option
will apply to columns that don't have a specific compression codec. See 
:class:`::arrow::Compression` for options.

Column data encodings can likewise be applied at the file-level or at the 

> column level. By default, the writer will attempt to dictionary encode all 

Reviewer 2: Perhaps
```suggestion
column level. By default, the writer will attempt to dictionary-encode all 
```

Author: Since it's serving as a verb I don't think it needs to be hyphenated.

supported columns, unless the dictionary grows too large. This behavior can
be changed at file-level or at the column level with ``disable_dictionary()``.
When not using dictionary encoding, it will fallback to the encoding set for 
the column or the overall file; by default ``Encoding::PLAIN``, but this can
be changed with ``encoding()``.

.. code-block:: cpp

   #include "parquet/arrow/writer.h"
   #include "arrow/util/type_fwd.h"

   using parquet::WriterProperties;
   using arrow::Compression;
   using parquet::Encoding;

   std::shared_ptr<WriterProperties> props = WriterProperties::Builder()
     .compression(Compression::SNAPPY)        // Fallback

>      ->compression("colA", Compression::ZSTD) // Only applies to colA

Reviewer 2: ```suggestion
     ->compression("colA", Compression::ZSTD) // Only applies to column "colA"
```

     ->encoding(Encoding::BIT_PACKED)         // Fallback
     ->encoding("colB", Encoding::RLE)        // Only applies to colB
     ->disable_dictionary("colB")             // Always use RLE, never dictionary

>      ->disable_statistics("colB")             // Stats won't be written for colB

Reviewer 2: Is it useful to show how not to write statistics?

     ->build();

Statistics are enabled by default for all columns. You can disable statistics for
all columns or specific columns using ``disable_statistics`` on the builder.
There is a ``max_statistics_size`` which limits the maximum number of bytes that
may be used for min and max values, useful for types like strings or binary blobs.

There are also Arrow-specific settings that can be configured with
:class:`parquet::ArrowWriterProperties`:

.. code-block:: cpp

   #include "parquet/arrow/writer.h"

   using parquet::ArrowWriterProperties;

   std::shared_ptr<ArrowWriterProperties> arrow_props = ArrowWriterProperties::Builder()
      .enable_deprecated_int96_timestamps() // default False
      ->store_schema() // default False
      ->enable_compliant_nested_types() // default False
      ->build();

These options mostly dictate how Arrow types are converted to Parquet types.
Turning on ``store_schema`` will cause the writer to place the serialized Arrow

> schema within the file metadata. This allows the Arrow reader to automatically
> determine which columns should be read back as dictionary-encoded columns,
> potentially saving memory.

Reviewer 2: Not only dictionary encoding is concerned. You can take a look at the existing rules [here](https://github.com/apache/arrow/blob/a9d2504b02f7c40a6c2dbed2a69ab6c447c1fa5b/cpp/src/parquet/arrow/schema.cc#L827-L996).

```suggestion
schema within the file metadata. Since there is no bijection between Parquet
schemas and Arrow schemas, storing the Arrow schema allows the Arrow reader
to more faithfully recreate the original data.
```

Author: Thanks! That's a really good point. I'll add that change and provide a few helpful examples of what this mapping does.

C++ implementation is part of the Apache Arrow project and benefits
from tight integration with the Arrow C++ classes and facilities.

Reading Parquet files
=====================

The :class:`arrow::FileReader` class reads data into Arrow Tables and Record
Batches.

> The :class:`StreamReader` and :class:`StreamWriter` classes allow for

Reviewer 2: I second this comment :-)

> The :class:`StreamReader` and :class:`StreamWriter` classes allow for

Reviewer 2: @wjones127 I think you forgot about this comment?

> The :class:`StreamReader` and :class:`StreamWriter` classes allow for

Author: Yes, sorry. Just pushed an update to fix this.

   :language: cpp
   :start-after: arrow::Status WriteFullFile(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 19-21
   :dedent: 2

.. note::

   Column compression is off by default in C++. See :ref:`below <parquet-writer-properties>` 
   for how to choose a compression codec in the writer properties.

To write out data batch-by-batch, use :class:`arrow::FileWriter`.

.. literalinclude:: ../../../cpp/examples/arrow/parquet_read_write.cc
   :language: cpp
   :start-after: arrow::Status WriteInBatches(
   :end-before: return arrow::Status::OK();
   :emphasize-lines: 23-25,32,36
   :dedent: 2

StreamWriter
------------

The :class:`StreamWriter` allows for Parquet files to be written using
standard C++ output operators.  This type-safe approach also ensures
that rows are written without omitting fields and allows for new row
groups to be created automatically (after certain volume of data) or
explicitly by using the :type:`EndRowGroup` stream modifier.

Exceptions are used to signal errors.  A :class:`ParquetException` is
thrown in the following circumstances:

* Attempt to write a field using an incorrect type.

> * Attempt to write too many fields in a row.

Reviewer 2: Too few as well, or are the remaining fields treated as omitted?

Author: Yes, it appears you can "skip" fields. Though this is part of the pre-existing docs that I moved around, so I didn't write this particular section.

There are also Arrow-specific settings that can be configured with
:class:`parquet::ArrowWriterProperties`:

.. code-block:: cpp

   #include "parquet/arrow/writer.h"

   using parquet::ArrowWriterProperties;

   std::shared_ptr<ArrowWriterProperties> arrow_props = ArrowWriterProperties::Builder()
      .enable_deprecated_int96_timestamps() // default False
      ->store_schema() // default False
      ->enable_compliant_nested_types() // default False
      ->build();

These options mostly dictate how Arrow types are converted to Parquet types.
Turning on ``store_schema`` will cause the writer to store the serialized Arrow
schema within the file metadata. Since there is no bijection between Parquet
schemas and Arrow schemas, storing the Arrow schema allows the Arrow reader
to more faithfully recreate the original data. This mapping from Parquet types
back to original Arrow types includes:

* Reading timestamps with original timezone information (Parquet does not
  support time zones);
* Reading Arrow types from their storage types (such as Duration from int64
  columns);
* Reading string and binary columns back into large variants with 64-bit offsets;

> * Reading back columns as dictionary encoded (whether an Arrow column and a 
>   the serialized Parquet version are dictionary encoded are independent).

Reviewer 2: ```suggestion
* Reading back columns as dictionary encoded (whether an Arrow column and
  the serialized Parquet version are dictionary encoded are independent).
```
